swap/peer.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2019 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package swap
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethersphere/swarm/p2p/protocols"
+)
+
+var SigDoesNotMatch = errors.New("signature does not match")
+var DontOwe = errors.New("no negative balance")
+
+// Peer is a devp2p peer for the Swap protocol
+type Peer struct {
+	*protocols.Peer
+	swap *Swap
+}
+
+func NewPeer(p *protocols.Peer, s *Swap) *Peer {
+	return &Peer{
+		Peer: p,
+		swap: s,
+	}
+}
+
+// handleMsg is for handling messages when receiving messages
+func (sp *Peer) handleMsg(ctx context.Context, msg interface{}) error {
+	switch msg := msg.(type) {
+
+	case *ChequeRequestMsg:
+		return sp.handleChequeRequestMsg(ctx, msg)
+
+	case *EmitChequeMsg:
+		return sp.handleEmitChequeMsg(ctx, msg)
+
+	case *ErrorMsg:
+		return sp.handleErrorMsg(ctx, msg)
+
+	case *ConfirmMsg:
+		return sp.handleConfirmMsg(ctx, msg)
+	}
+
+	return nil
+}
+
+// handleChequeRequestMsg runs when a peer receives a `ChequeRequestMsg`
+// It is thus run by the debitor
+// So the debitor needs to:
+//   * check that it indeed owes to the requestor (if not, ignore message)
+//   * check serial number
+//   * check amount
+//   * if all is ok, issue the cheque
+func (sp *Peer) handleChequeRequestMsg(ctx context.Context, msg interface{}) (err error) {
+	// check we have indeed a negative balance with the peer
+	var req *ChequeRequestMsg
+	var ok bool
+	var peerBalance int64
+
+	if req, ok = msg.(*ChequeRequestMsg); !ok {
+		return fmt.Errorf("Unexpected message type: %v", err)
+	}
+
+	peer := req.Peer
+
+	sp.swap.lock.Lock()
+	defer sp.swap.lock.Unlock() //TODO: Do we really want to block so long?
+
+	if peerBalance, ok = sp.swap.balances[peer]; !ok {
+		return fmt.Errorf("No exchanges with peer: %v", peer)
+	}
+	// do we actually owe to the remote peer?
+	if peerBalance >= 0 {
+		// TODO: should we send a message to the peer?
+		return DontOwe
+	}
+
+	// balance is negative, send a cheque
+	// TODO: merge with thresholds; need to check for threshold?
+	// if not, any negative balance will result in a cheque at this point
```

Comment:
I think this is a very important question.

I'm not sure we should check for a threshold here. Balances can be different between peers, and at some point thresholds as well, so even if the balance is below a threshold, the request for a check could be legitimate.

To add to your comment: a cheque will be issued when having any negative balance _only if_ a cheque request is received. 

We can expect well-behaved nodes to send cheque requests only when the balance goes over the threshold, in their favor.

If a malicious node issues a cheque requests while under the threshold (meaning: too soon), is there any harm to responding with a cheque as long as the issuing node agrees that a balance is in their peer's favor? This is already checked by the balance being lower than 0.

Thoughts?

Reply:
Well this is tricky. I suggest to discuss with research (@nagydani, @zelig).

The reason is I think we should not allow any request to result in a cheque if the balance is negative. It could encourage some spamming and creates unnecessary overhead. The accounting has been defined as a "floating adjustment" between positive and negative threshold, it would be bad IMHO if every negative tilt could result in a cheque.

I think this should be discussed together with the bigger question of how to cater for dynamic or non-matching thresholds.

Reply:
I will make a note of this as well. The problem is that there could be eventual differences in balances which might make it harder for both nodes to agree on when a cheque request is valid.

The other side of the coin is what you mention: spamming.